Automattic/vip-go-mu-plugins
# Search: Better request handling and caching

## Description

Currently, we don't really cache anything at all, this puts extra strain on the cluster.

This PR introduces low-level caching - we're caching the whole response object. The significant benefit of that is that any other logic would treat is a regular request. E.g. the query will appear in Search Dev Tools.

Currently, three following endpoints are cached `_doc`, `_mget`, `_search`.

Cache time is set at 5 minutes (for fallback purposes), but a request is deemed stale at 30s mark. This allows for content freshness + graceful fallback.


Introduces filters:

```apply_filters( 'vip_search_cache_es_response', bool $is_cacheable, string $url, array $args )```

```apply_filters( 'vip_search_stale_request_threshold', 30, $args );```

## Changelog Description

### Plugin Updated: Enterprise Search

We have introduced a graceful fallback in case of a request to Enterprise Search backend fails. We would serve a previously-stored stale response (up to 5 minutes).

Additionally, only `query` requests are counted against the fail counter that blocks subsequent remote requests in case of a failure (via vip_safe_wp_remote_request).

## Checklist

Please make sure the items below have been covered before requesting a review:


## Steps to Test

1. Check out the PR
2. Set a breakpoint or  add an exit inside the https://github.com/Automattic/vip-go-mu-plugins/pull/2597/files#diff-dd3f02b793bcc2930a3ba1d70b0520c0865dc4a09d2dbd1f49a1c55af101c40dR853 block
3. Hit a `?s=test` or similar URL in the browser
4. Verify that. on the second run the cached code path is executing
